From 33f89ad9e564eda2d2b008fca955b8827e25cc4d Mon Sep 17 00:00:00 2001
From: Sanchayan Maity <sanchayan@centricular.com>
Date: Wed, 20 May 2026 20:06:18 +0530
Subject: [PATCH 16/18] openssl: Do not hold mutex across blocking I/O wait

This also fixes the test_simultaneous_sync timing out when running the
connection-openssl tests.
---
 tls/openssl/gtlsconnection-openssl.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/tls/openssl/gtlsconnection-openssl.c b/tls/openssl/gtlsconnection-openssl.c
index 83e9168..a8b8bcc 100644
--- a/tls/openssl/gtlsconnection-openssl.c
+++ b/tls/openssl/gtlsconnection-openssl.c
@@ -337,7 +337,9 @@ perform_openssl_io (GTlsConnectionOpenssl  *openssl,
       if (io_timeout == 0)
         break;
 
+      g_tls_connection_openssl_unlock (openssl);
       g_tls_bio_wait_available (priv->bio, io_needed, io_timeout, cancellable);
+      ssl = g_tls_connection_openssl_lock_ssl (openssl);
     }
 
   if (status == G_TLS_CONNECTION_BASE_TRY_AGAIN)
-- 
2.20.1.windows.1

